ARCHITECTURE.md

```diff
@@ -26,8 +30,41 @@ The graphql server can scale horizontally with additional pods.
 
 Redis is being used as a distributed lock such that several accounts can't be modified at the same time by different instances. 
 
-### Database
+## Database
+
+The state of the wallet is shared among 4 differences sources: 
+
+### bitcoin-core
+
+bitcoin core handle the cold storage. 
+
+there is typically a very low load transaction wise in regard to the wallet. maybe 1 transaction per week when a rebalancing is needed.
+
+### lnd 
+
+lnd handle all of the transactions the users run.
+lnd stores itself the data in a bbolt database.
+
+every succesful payment within lnd is being recorded on mongodb.
+
+offchain-wise, nothing is needed within the lnd side to present a transaction to a user after it has been executed.
+onchain-wise, it's possible that a user shared an onchain address. this is currently tied to lnd. (we intend to not use lnd for onchain transaction to remove this dependancy and simplify lnd instance recyling)
+
+### Mongodb
+
+Mongodb is the database storing all users transactions. It's the source of truth for the customers' accounts.
+
+To achieve strong consistency, it's currently setup with:
+- Write access need to be valided by 2-out-of-3 pods before being considered succesful
```

Comment:
I have been reading a bit in the MongoDB docs to see what consistency guarantees there are. It appears with Majority write concern (as is being used in the deploy scripts) we are indeed guaranteed that the write only succeeds in a replica set once a majority of nodes have persisted the data. (https://docs.mongodb.com/manual/reference/write-concern/)

The problem as far as I can tell is that reads have no such guarantee. By default the read concern is 'local' which will only query the primary or 'available' if querying a secondary. Only with read concern 'linearisable' is there guarantee of consistent reads where all data that is being read has the guarantee of being written to a majority of nodes. https://docs.mongodb.com/manual/reference/read-concern-linearizable/ .

I cannot currently see any place in the code base where this is configured so my assumption is we aren't using it. I also do not have enough awareness of the exact order of processing in the code to asses what inconsistencies we risk via this behaviour. But it doesn't make me feel confident about things working correctly at scale.